runtime/doc/news.txt

```diff
@@ -73,7 +73,12 @@ The following changes may require adaptations in user config or plugins.
 
 • |OptionSet| autocommand args |v:option_new|, |v:option_old|,
   |v:option_oldlocal|, |v:option_oldglobal| now have the type of the option
-  instead of always being strings.
+  instead of always being strings. |v:option_old| is now the old global value
+  for all global-local options, instead of just string global-local options.
+
+• Local value for a global-local number/boolean option is now unset when
+  the option is set (e.g. using |:set| or |nvim_set_option_value()|) without a
```

Comment:
I assume this is only the case when buf/win is passed to `nvim_set_option_value`?

Reply:
> I assume this is only the case when buf/win is passed to `nvim_set_option_value`?

Nope

Reply:
> So when we use `nvim_set_option_value` with no scope it'll reset the local values for all buffers/windows?

No, only the current buffer/window. And it already did that for string options, btw.

Reply:
This seems undesirable, at least for the API and should probably be documented. The API should be agnostic to current win/buf.

Reply:
> This seems undesirable, at least for the API and should probably be documented. The API should be agnostic to current win/buf.

Again, this was already the case for string options. The only thing this PR does is make it more consistent so that it also occurs for number and boolean options. I would be fine with documenting this, though.

> The API should be agnostic to current win/buf.

I generally agree, but unfortunately this was not how options were designed to work. If we really want we could probably add some extra code to `nvim_set_option_value` to set the local value for the curbuf again but I'm not sure if we really want that. I'll defer this to @bfredl as I don't have any strong opinions on this.

Reply:
Also, using `:set` (or `nvim_set_option_value()`)  without scope implies setting both the local and global values for a global-local option. All this does is add an optimization (for non-trivial types) by removing the local value since it won't be used anyway.

Reply:
> Again, this was already the case for string options. 

I know! I didn't ignore that.

> The only thing this PR does is make it more consistent so that it also occurs for number and boolean options. I would be fine with documenting this, though.

By making it consistent you've had to change behaviour for a certain type of options. You could have just as easily changed how strings work. 

Either way it seems like we may want to change behaviour in the future so need to be weary of flip flopping.

Reply:
Potentially. I haven't got all the necessary context in my brain atm, just that a design goal of the API is to also pass win/buf when it acts on them and pass 0 for current win/buf. Deviating from that would be unfortunate.

Reply:
This will be resolved in a follow-up PR